095-stretch-cluster.md

```diff
@@ -0,0 +1,318 @@
+
+```
+advertised.listeners=REPLICATION-9091://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9091,
+PLAIN-9092://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9092,
+TLS-9093://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9093
+```
+
+**Modified format for a stretch Kafka cluster**
+
+When a Kafka cluster is deployed across multiple Kubernetes clusters, the operator modifies `advertised.listeners` to include a Kubernetes cluster identifier (`stretch-cluster-id`) that the user defines when [configuring a supported Cloud Native Network project](#step-1-user-configuration-of-a-supported-cloud-native-network-project).
+The modified format is:
+
+```
+advertised.listeners=REPLICATION-9091://<broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9091,
+PLAIN-9092://<broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9092,
+TLS-9093://<broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9093
+```
+
+The `<stretch-cluster-id>` is dynamically retrieved from an [annotation](#kafkanodepool-cr) on the `KafkaNodePool` resource.
+
+**Modified `controller.quorum.voters` format**
+
+Similarly, `controller.quorum.voters` includes all Kafka controllers running in all of the Kubernetes clusters involved in the stretch Kafka cluster:
+
+```
+controller.quorum.voters=<controller-id>@<controller-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9090, ...
+```
+
+These updates ensure brokers and controllers can be discovered and communicate across clusters without relying on traditional external access methods.
+
+#### Resource cleanup on remote Kubernetes clusters
+
+In our POC, resources created in the remote clusters do not have `OwnerReferences`. The main reason is that the `Kafka` and `KafkaNodePool` CRs exist only in the central cluster. If resources in remote clusters were to reference them as owners, they would be immediately deleted by Kubernetes' garbage collection mechanism since their owners do not exist in the remote clusters.
+
+Even if ownership across Kubernetes cluster boundaries were possible, it would introduce another issue: if the central cluster were to go down, any resources in remote clusters with OwnerReferences pointing to central cluster resources would also be deleted automatically. This is not the intended behaviour, as we want remote resources to remain operational even if the central cluster becomes temporarily unavailable.
+
+The prototype does not yet have a mechanism for controlled cleanup of remote resources when a user deliberately deletes the Kafka and KafkaNodePool CRs from the central cluster. Currently, when a user deletes these CRs, all related resources in the central cluster are removed, but remote cluster resources remain. Finding a way to handle this cleanup while avoiding unintended deletions is an open challenge.
```

Comment:
Why doesn't the central cluster delete the corresponding resources on the remote clusters as it creates them in the beginning?

Reply:
We are currently working on this area and hope to have some updates next week to help answer your question and clarify the proposal here.

Reply:
We have updated this section with a more precise proposal about how remote resources will be managed.